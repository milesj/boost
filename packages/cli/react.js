module.exports = require('./cjs/react.cjs');
