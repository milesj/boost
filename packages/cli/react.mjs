export * from './mjs/react.mjs';
