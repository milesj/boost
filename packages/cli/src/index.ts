/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
