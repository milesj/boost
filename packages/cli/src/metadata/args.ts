export * from '../decorators';
export * from '../initializers';
