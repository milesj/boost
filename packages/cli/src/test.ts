/* eslint-disable no-param-reassign */

import React from 'react';
import { render } from 'ink';
import { env } from '@boost/internal';
import { mockLogger } from '@boost/log/test';
import { stripAnsi } from '@boost/terminal';
import { Command, INTERNAL_OPTIONS, INTERNAL_PARAMS, Program } from '.';

import type {
	ExitCode,
	GlobalOptions,
	PrimitiveType,
	ProgramOptions,
	ProgramStreams,
	TaskContext,
} from '.';

export class MockReadStream {
	isTTY = false;
}

export class MockWriteStream {
	append: boolean = false;

	columns: number = 80;

	output: string;

	constructor(append: boolean = false) {
		this.append = append;
		this.output = '';
	}

	write(string: string) {
		if (this.append) {
			this.output += string;
		} else {
			this.output = string;
		}
	}

	get(): string {
		return this.output;
	}

	on() {}

	off() {}
}

/**
 * Returns mocked `stderr`, `stdout`, and `stdin` streams that can be passed to a `Program`.
 * This does not mock all stream functionality, only those required by Boost and Ink.
 *
 * ```ts
 * import { mockStreams } from '@boost/cli/test';
 *
 * const streams = mockStreams();
 * ```
 */
export function mockStreams(append?: boolean): ProgramStreams {
	return {
		stderr: new MockWriteStream(append),
		stdin: new MockReadStream(),
		stdout: new MockWriteStream(append),
	} as unknown as ProgramStreams;
}

/**
 * Returns a `Program` instance with required options pre-filled and streams mocked
 * (unless manually provided).
 *
 * ```ts
 * import { mockProgram } from '@boost/cli/test';
 *
 * const program = mockProgram({ name: 'Example' });
 * ```
 */
export function mockProgram(options?: Partial<ProgramOptions>, streams?: ProgramStreams): Program {
	return new Program(
		{
			bin: 'test',
			name: 'Test',
			version: '0.0.0',
			...options,
		},
		streams ?? mockStreams(),
	);
}

/**
 * Can be used to render a React component with Ink and return the rendered result
 * as a terminal compatible string. If `stripped` is true, it will strip ANSI
 * escape escape sequences.
 *
 * ```tsx
 * import { renderComponent } from '@boost/cli/test';
 * import TestComponent from '../src/components/TestComponent';
 *
 * it('renders a component', async () => {
 * 	expect(await renderComponent(<TestComponent />)).toMatchSnapshot();
 * });
 * ```
 *
 * > As an alternative, we also suggest using the official
 * > [ink-testing-library](https://github.com/vadimdemedes/ink-testing-library).
 */
export async function renderComponent(
	element: React.ReactElement,
	stripped: boolean = false,
): Promise<string> {
	let output = '';

	try {
		const { render: renderAsTest } = await import('ink-testing-library');
		const { lastFrame } = await renderAsTest(element);

		output = lastFrame() ?? '';
	} catch {
		const stdout = new MockWriteStream();

		await render(element, {
			debug: true,
			experimental: true,
			stdout: stdout as unknown as NodeJS.WriteStream,
		});

		output = stdout.get();
	}

	return stripped ? stripAnsi(output) : output;
}

/**
 * Runs a `Command` outside the context of a `Program`, but mimics similar functionality,
 * including React component rendering. Params are required as they're passed to the run method,
 * while options are optional and assume class properties have been defined. Also, the `exit`
 * and `log` methods have been mocked with Vitest spies so that they can be asserted.
 *
 * ```ts
 * import { runCommand } from '@boost/cli/test';
 * import TestCommand from '../src/commands/TestCommand';
 *
 * it('runs a command', async () => {
 * 	const command = new TestCommand();
 *
 * 	expect(await runCommand(command, ['foo', 'bar', 'baz'])).toMatchSnapshot();
 * 	expect(command.log).toHaveBeenCalled();
 * });
 * ```
 *
 * > Since there is no `Program` context, any functionality that requires a program will fail. If
 * > needed, use `runProgram()` instead.
 */
export async function runCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
	command: Command<O, P>,
	params: P,
	options?: Partial<O>,
): Promise<string> {
	const { vi } = await import('vitest');

	if (options) {
		Object.assign(command, options);

		// @ts-expect-error Allow overwrite
		command[INTERNAL_OPTIONS] = {
			help: false,
			locale: 'en',
			version: false,
			...options,
		};
	}

	command.exit = vi.fn();
	// eslint-disable-next-line require-atomic-updates
	command.log = await mockLogger();
	// eslint-disable-next-line require-atomic-updates
	command[INTERNAL_PARAMS] = params;

	const result = await command.run(...params);

	if (!result || typeof result === 'string') {
		return result ?? '';
	}

	return renderComponent(result);
}

/**
 * Runs a task function outside the context of a `Command`, in complete isolation.
 * A mock command context is provided with standard defaults, and can be customized
 * through the 3rd argument.
 *
 * ```ts
 * import { runTask } from '@boost/cli/test';
 * import testTask from '../src/tasks/testTask';
 *
 * it('runs a task', async () => {
 * 	const context = {
 * 		log: vi.fn(),
 * 	};
 *
 * 	expect(await runTask(testTask, ['foo', 'bar', 'baz'], context)).toMatchSnapshot();
 * 	expect(context.log).toHaveBeenCalled();
 * });
 * ```
 */
export async function runTask<A extends unknown[], R, T extends TaskContext>(
	task: (this: T, ...argz: A) => R,
	args: A,
	context?: Partial<T>,
): Promise<R> {
	const { vi } = await import('vitest');
	const notTestable = (name: string) => () => {
		throw new Error(
			`\`${name}\` is not testable using the \`runTask\` utility. Test using a full program.`,
		);
	};

	const baseContext: TaskContext = {
		exit: vi.fn(),
		help: false,
		locale: 'en',
		log: await mockLogger(),
		rest: [] as string[],
		runProgram: notTestable('runProgram'),
		runTask: notTestable('runTask'),
		unknown: {},
		version: false,
	};

	return task.apply(
		{
			...baseContext,
			...context,
		} as T,
		args,
	);
}

/**
 * Runs a `Program` as if it were ran on the command line, including middleware,
 * commands, success and failure states, and more. Utilizes mocked streams to capture
 * and return standard output and ANSI stripped output. Failed runs will not throw
 * and instead will render a failure output.
 *
 * ```ts
 * import { runProgram } from '@boost/cli/test';
 * import Program from '../src/program';
 *
 * it('runs a program', async () => {
 * 	const program = new Program();
 *
 * 	const { code, output } = await runProgram(program, ['cmd', '--foo', '123', 'bar']);
 *
 * 	expect(output).toMatchSnapshot();
 * 	expect(code).toBe(0);
 * });
 * ```
 */
export async function runProgram(
	program: Program,
	argv: string[],
	options: { append?: boolean } = {},
): Promise<{ code: ExitCode; output: string; outputStripped: string }> {
	if (!(program.streams.stderr instanceof MockWriteStream)) {
		program.streams.stderr = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdout instanceof MockWriteStream)) {
		program.streams.stdout = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdin instanceof MockReadStream)) {
		program.streams.stdin = new MockReadStream() as unknown as NodeJS.ReadStream;
	}

	// Ink async rendering never resolves while testing,
	// as it relies on system signals to "exit".
	// So we set this to flag the renderer to avoid awaiting.
	env('CLI_TEST_ONLY', 'true');

	const code = await program.run(argv);

	env('CLI_TEST_ONLY', null);

	const output =
		(program.streams.stdout as unknown as MockWriteStream).get() +
		(program.streams.stderr as unknown as MockWriteStream).get();

	return {
		code,
		output,
		outputStripped: stripAnsi(output),
	};
}
