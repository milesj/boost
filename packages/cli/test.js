const { Command } = require('./lib');

const c = new Command();
