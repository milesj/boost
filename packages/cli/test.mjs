export * from './mjs/test.mjs';
