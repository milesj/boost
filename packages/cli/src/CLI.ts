import { Argv } from '@boost/args';
import { Contract, Predicates } from '@boost/common';
import { CLIOptions } from './types';

export default class CLI extends Contract<CLIOptions> {
  blueprint({ string }: Predicates) {
    return {
      bin: string().required(),
      name: string().required(),
      version: string().required(),
    };
  }

  run(argv: Argv) {}
}
