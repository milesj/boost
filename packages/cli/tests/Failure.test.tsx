import React from 'react';
import { ParseError, ValidationError } from '@boost/args';
import { jest } from '@jest/globals';
import { Failure } from '../src/react';
import { renderComponent } from '../src/test';

jest.mock('term-size');

describe('<Failure />', () => {
	it('renders a common error', async () => {
		expect(
			await renderComponent(<Failure binName="boost" error={new Error('Something is broken!')} />),
		).toMatchSnapshot();
	});

	it('doesnt render code frame if a common error', async () => {
		expect(
			await renderComponent(
				<Failure
					binName="boost"
					commandLine="foo --bar"
					error={new Error('Something is broken!')}
				/>,
			),
		).toMatchSnapshot();
	});

	it('doesnt render code frame if arg not found', async () => {
		const error = new ParseError(
			'Flags and short option groups may not use inline values.',
			'--flag=123',
			16,
		);

		expect(
			await renderComponent(<Failure binName="boost" commandLine="foo --bar" error={error} />),
		).toMatchSnapshot();
	});

	it('renders an error with warnings', async () => {
		const warnings = [
			new Error('This is invalid.'),
			new Error('This is also invalid. Please fix.'),
		];

		const out = await renderComponent(
			<Failure binName="boost" error={new Error('Something is broken!')} warnings={warnings} />,
		);

		// Do not use snapshots a they differ on windows
		expect(out).toContain('This is invalid.');
		expect(out).toContain('This is also invalid. Please fix.');
	});

	it('renders a parse error', async () => {
		const error = new ParseError(
			'Flags and short option groups may not use inline values.',
			'--flag=123',
			16,
		);

		expect(await renderComponent(<Failure binName="boost" error={error} />)).toMatchSnapshot();
	});

	it('renders a parse error with a command line', async () => {
		const error = new ParseError(
			'Flags and short option groups may not use inline values.',
			'--flag=123',
			16,
		);

		expect(
			await renderComponent(
				<Failure binName="boost" commandLine="--foo value --flag=123 -gSA" error={error} />,
				true,
			),
		).toMatchSnapshot();
	});

	it('renders a validation error', async () => {
		const error = new ValidationError('Not enough arity arguments.', 'foo');

		expect(await renderComponent(<Failure binName="boost" error={error} />)).toMatchSnapshot();
	});

	it('renders a validation error with a command line', async () => {
		const error = new ValidationError('Not enough arity arguments.', 'foo');

		expect(
			await renderComponent(
				<Failure binName="boost" commandLine="--foo value --flag=123 -gSA" error={error} />,
				true,
			),
		).toMatchSnapshot();
	});

	it('reduces command line by half when index appears off screen', async () => {
		const line = `--foo value --bar "${'w'.repeat(100)}" --flag=123 -gSA "${'m'.repeat(75)}"`;

		const error = new ParseError(
			'Flags and short option groups may not use inline values.',
			'--flag=123',
			line.indexOf('--flag=123'),
		);

		expect(
			await renderComponent(<Failure binName="boost" commandLine={line} error={error} />, true),
		).toMatchSnapshot();
	});
});
