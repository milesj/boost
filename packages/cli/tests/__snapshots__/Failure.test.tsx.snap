// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Failure /> doesnt render code frame if a common error 1`] = `
"
[30m[41m ERROR [49m[39m

  Something is broken!
"
`;

exports[`<Failure /> reduces command line by half when index appears off screen 1`] = `
"
 ERROR 

  Flags and short option groups may not use inline values.

                                               ┌────────┐
  … wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww\\" --flag=123 -gSA \\"mmmmmmmmmmmmmmmm …
"
`;

exports[`<Failure /> renders a common error 1`] = `
"
[30m[41m ERROR [49m[39m

  Something is broken!
"
`;

exports[`<Failure /> renders a parse error 1`] = `
"
[30m[41m ERROR [49m[39m

  Flags and short option groups may not use inline values.
"
`;

exports[`<Failure /> renders a parse error with a command line 1`] = `
"
 ERROR 

  Flags and short option groups may not use inline values.

                  ┌────────┐
  bin --foo value --flag=123 -gSA
"
`;

exports[`<Failure /> renders a validation error 1`] = `
"
[30m[41m ERROR [49m[39m

  Not enough arity arguments.
"
`;

exports[`<Failure /> renders a validation error with a command line 1`] = `
"
 ERROR 

  Not enough arity arguments.

      ┌───┐
  bin --foo value --flag=123 -gSA
"
`;

exports[`<Failure /> renders an error with warnings 1`] = `
"
[30m[41m ERROR [49m[39m

  Something is broken!

[30m[43m WARNINGS [49m[39m

  ● This is invalid.
  ● This is also invalid. Please fix.
"
`;
